plugins/pixart-tp/fu-pxi-tp-firmware.c

```diff
@@ -0,0 +1,953 @@
+	if (fw == NULL)
+		return NULL;
+	fu_byte_array_append_bytes(buf, fw);
+	return g_steal_pointer(&buf);
+}
+
+/**
+ * fu_pxi_tp_firmware_build:
+ * @firmware: the container object
+ * @n: parsed builder XML node (unused)
+ * @error: return location for a #GError, or %NULL
+ *
+ * Optional builder hook for constructing a container from a .builder.xml.
+ * Not used for the PixArt FWHD v1.0 parser; present to satisfy the vfunc.
+ *
+ * Returns: %TRUE always.
+ */
+static gboolean
+fu_pxi_tp_firmware_build(FuFirmware *firmware, XbNode *n, GError **error)
+{
+	(void)firmware;
+	(void)n;
+	(void)error; /* parser-only */
+	return TRUE;
+}
+
+static void
+fu_pxi_tp_firmware_init(FuPxiTpFirmware *self)
+{
+	self->sections = g_ptr_array_new_with_free_func(g_free);
+	fu_firmware_add_flag(FU_FIRMWARE(self), FU_FIRMWARE_FLAG_HAS_CHECKSUM);
+}
+
+/* ---------------- finalize ---------------- */
+static void
+fu_pxi_tp_firmware_finalize(GObject *obj)
+{
+	FuPxiTpFirmware *self = FU_PXI_TP_FIRMWARE(obj);
+	if (self->sections != NULL) {
+		g_ptr_array_unref(self->sections); /* safe: has free_func=g_free */
+		self->sections = NULL;
+	}
+	G_OBJECT_CLASS(fu_pxi_tp_firmware_parent_class)->finalize(obj);
+}
+
+static void
+fu_pxi_tp_firmware_class_init(FuPxiTpFirmwareClass *klass)
+{
+	GObjectClass *object_class = G_OBJECT_CLASS(klass);
+	FuFirmwareClass *firmware_class = FU_FIRMWARE_CLASS(klass);
+
+	object_class->finalize = fu_pxi_tp_firmware_finalize;
+
+	firmware_class->validate = fu_pxi_tp_firmware_validate;
+	firmware_class->parse = fu_pxi_tp_firmware_parse;
+	firmware_class->export = fu_pxi_tp_firmware_export;
+	firmware_class->write = fu_pxi_tp_firmware_write;
+	firmware_class->build = fu_pxi_tp_firmware_build;
+}
+
+/* ---------------- factory ----------------- */
+FuFirmware *
+fu_pxi_tp_firmware_new(void)
+{
+	return FU_FIRMWARE(g_object_new(FU_TYPE_PXI_TP_FIRMWARE, NULL));
+}
+
+/* -------------------------- getters ----------------------- */
+guint16
+fu_pxi_tp_firmware_get_header_version(FuPxiTpFirmware *self)
+{
+	g_return_val_if_fail(FU_IS_PXI_TP_FIRMWARE(self), 0);
+	return self->header_ver;
+}
+
+guint16
+fu_pxi_tp_firmware_get_file_version(FuPxiTpFirmware *self)
+{
+	g_return_val_if_fail(FU_IS_PXI_TP_FIRMWARE(self), 0);
+	return self->file_ver;
+}
+
+guint16
+fu_pxi_tp_firmware_get_ic_part_id(FuPxiTpFirmware *self)
+{
+	g_return_val_if_fail(FU_IS_PXI_TP_FIRMWARE(self), 0);
+	return self->ic_part_id;
+}
+
+guint16
+fu_pxi_tp_firmware_get_total_flash_sectors(FuPxiTpFirmware *self)
+{
+	g_return_val_if_fail(FU_IS_PXI_TP_FIRMWARE(self), 0);
+	return self->flash_sectors;
+}
+
+guint16
+fu_pxi_tp_firmware_get_num_valid_sections(FuPxiTpFirmware *self)
+{
+	g_return_val_if_fail(FU_IS_PXI_TP_FIRMWARE(self), 0);
+	return self->num_sections;
+}
+
+const GPtrArray *
+fu_pxi_tp_firmware_get_sections(FuPxiTpFirmware *self)
+{
+	g_return_val_if_fail(FU_IS_PXI_TP_FIRMWARE(self), NULL);
+	return self->sections;
+}
```

Comment:
I think you're reinvented `fu_firmware_get_images()` :) -- I think each section should be a "child" -- e.g. use 

    FuFirmware *img = fu_firmware_new()
    if (!fu_firmware_parse(img, the_section_bytes, error))
        return FALSE;
    fu_firmware_add_image(firmware, img);

when you're parsing the sections...